Additional ID Token claim checks on OAUTH2 authorization

EvertonSA opened the pull request:
I'm opening this MR again and I'm whiling to resolve all threads comments. 

----

~~This MR introduces possibility to lock your pgadmin4 web deployment leveraging GitLab Groups and Roles.~~

~~4 new fields added to oauth2_config object:~~

~~OAUTH_GITLAB_AUTHZ_ENABLED: feature flag~~
~~OAUTH_GITLAB_GROUPS_OWNER_ROLE: list of authorized owners group on GitLab~~
~~OAUTH_GITLAB_GROUPS_MAINTAINER_ROLE: list of authorized maintainers group on GitLab~~
~~OAUTH_GITLAB_GROUPS_DEVELOPER_ROLE: list of authorized developers group on GitLab~~


~~at this moment, the ROLE variables are not actually a list, but a string separated by space "group1 group2 group3"~~

This MR introduces the possibility of checking additional claims on the ID Token given by the OAUTH2 provider. Any match is sufficient for authorized access, and if there is no match, the login flow is canceled with a reasonable error message.

adityatoshniwal commented:
@EvertonSA this looks like adding native support for gitlab to pgAdmin. Unfortunately, this cannot be accepted until it is made to work using configuration, and without using the name `gitlab` anywhere.

EvertonSA commented:
GitHub uses concept of Organization, and GitLab uses the concept of Groups. 

It needs to be GitLab specific. This setup is using config, not environment variables. This config goes into the OAUTH.json.

How would you approach that to: "without using the name gitlab anywhere."?

dpage commented:
I'm not familiar with the design of OAuth, but if this is a standard or common way of implementing authz, then Aditya is correct. In that case we would use generic names and document the providers it works with (or definitely does not work with) and give some configuration examples.

For example:

```
OAUTH_AUTHZ_ENABLED: feature flag
OAUTH_OWNER_ROLE: list of authorized owners group on OAuth provider
OAUTH_MAINTAINER_ROLE: list of authorized maintainers group on OAuth provider
OAUTH_DEVELOPER_ROLE: list of authorized developers group on OAuth provider
```

(sidenote: it looks like you're expecting those settings to have a space-delimited list of values. The config file is in Python, so why not make them lists?)

If this truly is a feature that only makes any sense on Gitlab, then we'd accept the name in the options. Otherwise, we need to think about how this can be made to work with other providers as well; for example, does Github have the same concepts of owner, maintainer, and developer roles? If not, perhaps we would provide a list of claims (is that the correct term?) in one config option and state you must be a member of one of these, e.g.

OAUTH_AUTHZ_CLAIMS = ['https://gitlab.org/claims/groups/owner', 'https://gitlab.org/claims/groups/maintainer', 'https://gitlab.org/claims/groups/developer']
OAUTH_AUTHZ_GROUPS = ['group1', 'group2']

That would tell us the list of authorised groups, and where to look for them. 

Additionally; any PR will require a documentation update :-)

adityatoshniwal commented:
@EvertonSA You can probably have something like `OAUTH2_OTHER_INFO`

EvertonSA commented:
I understand everyone concern I'm aware my code sucks. 

I feel a little bumped I have to decipher github oauth2 to implement gitlab oauth2 here, but, i'm willing to.

Documentation is not an issue I will do, I just ask for implementation details guidance and patience.

keep this open I will work little by little

dpage commented:
> I understand everyone concern I'm aware my code sucks.

No-one said your code sucks! This is about thinking more broadly and coming up with a design that can be applied to multiple situations, so we don't end up with a mess of code and configuration options.

Scratch-your-own-itch coding is great to solve your own problem, but as maintainers of pgAdmin we have to think about the bigger picture that includes different use cases and code maintainability over the long term.

> I feel a little bumped I have to decipher github oauth2 to implement gitlab oauth2 here, but, i'm willing to.

I was hoping it's less about deciphering Github specifically, and more about the OAuth2 spec in general. I'm assuming that the way Gitlab tell us what groups a user is in is part of that spec, and not something specific to them. If it is in the spec, then we'd want a patch that can accommodate other implementations. If not, then that's where Gitlab-specific code could be considered.

> Documentation is not an issue I will do, I just ask for implementation details guidance and patience.
> 
> keep this open I will work little by little

No problem - thank you for your efforts. Don't be afraid to ask us for help or guidance if you need to; we're happy to assist.

EvertonSA commented:
Below is userinfo from a GitHub organization app VS GitLab group app

GitHub profile:

```
{
  "login": "EvertonSA",
  "id": 30481051,
  "node_id": "MDQREDACTUx",
  "avatar_url": "https://avatars.githubusercontent.com/u/30481051?v=4",
  "gravatar_id": "",
  "url": "https://api.github.com/users/EvertonSA",
  "html_url": "https://github.com/EvertonSA",
  "followers_url": "https://api.github.com/users/EvertonSA/followers",
  "following_url": "https://api.github.com/users/EvertonSA/following{/other_user}",
  "gists_url": "https://api.github.com/users/EvertonSA/gists{/gist_id}",
  "starred_url": "https://api.github.com/users/EvertonSA/starred{/owner}{/repo}",
  "subscriptions_url": "https://api.github.com/users/EvertonSA/subscriptions",
  "organizations_url": "https://api.github.com/users/EvertonSA/orgs",
  "repos_url": "https://api.github.com/users/EvertonSA/repos",
  "events_url": "https://api.github.com/users/EvertonSA/events{/privacy}",
  "received_events_url": "https://api.github.com/users/EvertonSA/received_events",
  "type": "User",
  "site_admin": false,
  "name": "Everton Seiei Arakaki",
  "company": null,
  "blog": "",
  "location": null,
  "email": null,
  "hireable": null,
  "bio": null,
  "twitter_username": null,
  "public_repos": 80,
  "public_gists": 2,
  "followers": 22,
  "following": 86,
  "created_at": "2017-07-26T21:55:22Z",
  "updated_at": "2023-02-22T22:10:40Z"
}
```

GitLab Profile:

```
{
  "sub": "4044311",
  "sub_legacy": "def2e18f5eaf9014dd7a2ea96dfe31703240e682b3990835c3be2519b6893b2b",
  "name": "Everton Arakaki",
  "nickname": "eveuca",
  "preferred_username": "eveuca",
  "email": "eveuca@gmail.com",
  "email_verified": true,
  "profile": "https://gitlab.com/eveuca",
  "picture": "https://secure.gravatar.com/avatar/3b71094e9c21812ffa50d24748e213ac?s=80&d=identicon",
  "groups": [
    "kuberheads",
    "kuberheads/applications",
    "kuberheads/platform"
  ],
  "https://gitlab.org/claims/groups/owner": [
    "kuberheads",
    "kuberheads/platform",
    "kuberheads/applications"
  ]
}
```

While GitLab provides `groups` and `"https://gitlab.org/claims/groups/owner"`,  `"https://gitlab.org/claims/groups/maintainer"`,  `"https://gitlab.org/claims/groups/developer"`, GitHub only gives you endpoints that you still need to HTTP fetch to perform further authorization logics.

I am still reluctant to accept this feature can be "generalized". I agree on @khushboovashi comments on `OAUTH2_CLAIMS: {'clam1': [list of values], 'claim2': [list of values], 'claim3': [list of values]}` but I don't know if someone will benefit from it other than GitLab implementation.

EvertonSA commented:
this is how grafana is checking if user is part of organization https://github.com/grafana/grafana/blob/f005a375db149be1981b8fad88a99d34c998cdc9/pkg/login/social/github_oauth.go#L67

From my humble understanding, this is another issue/PR.

EvertonSA commented:
I will be proceeding with refactoring and include in the docs that github organization check is not available.

EvertonSA commented:
relates to https://github.com/pgadmin-org/pgadmin4/issues/4467

EvertonSA commented:
@adityatoshniwal can you kindly review this change?

EvertonSA commented:
thanks for triggering the pipeline, I will fix the existing issues and ping again! THANKS

EvertonSA commented:
This should pass all pycodestyle errors, but I'm usure on the postgres 15 error.

I'm willing to resolve any other comments on this PR.

EvertonSA commented:
oh well great, more pycodestyle to be fixed, will fix by the end of the day

adityatoshniwal commented:
Hi @EvertonSA,

You can go to `web` directory and run `yarn run pep8` to check locally.

EvertonSA commented:
Thank you a lot for the patient here, I'm really grateful for that. 

will fix as soon as I get home

EvertonSA commented:
thanks for everyone patience here, hope it works now!

EvertonSA commented:
I think the problem now it's just flakeness on the psql tests, anything I can do to help?

khushboovashi commented on docs/en_US/oauth2.rst:
```diff
@@ -39,6 +43,10 @@ and modify the values for the following parameters:
     create a pgAdmin user corresponding to a successfully authenticated Oauth2 user.
     Please note that password is not stored in the pgAdmin database."
+    "OAUTH2_ADDITIONAL_CLAIMS", "The additional claims to check on the openid ID token. 
+    If the value is provided, and the user ID Token has any of the given values, the login flow will continue.
```
The description is not very clear.

EvertonSA replied:
I'm willing to change but i'm out of ideas, how would it look very clear to you?

EvertonSA replied:
for reference, this is how hashicorp vault implement a similar feature:

![image](https://github.com/pgadmin-org/pgadmin4/assets/30481051/59759472-5c3a-4269-a663-af169da2ae60)

EvertonSA replied:
changed, can you see if it looks better now?

khushboovashi commented on web/pgadmin/authenticate/oauth2.py:
```diff
@@ -151,6 +151,22 @@ def login(self, form):
                 current_app.logger.exception(error_msg)
                 return False, gettext(error_msg)
 
+        if self.oauth2_config[self.oauth2_current_client][
```
If OAUTH2_ADDITIONAL_CLAIMS is not specified (even as None) in the config, this code will return an error. Make sure it exists in the array, and proceed if it exists.

Add test cases for this config. You can modify the test_oauth2_with_mocking.py file to add the test cases.

EvertonSA replied:
will do

EvertonSA replied:
added validation for success check on OAUTH2_ADDITIONAL_CLAIMS, but I don't know how to assert a failure and I could not find examples. If this is mandatory, can you help me here?

khushboovashi commented on web/pgadmin/authenticate/oauth2.py:
```diff
@@ -151,6 +151,22 @@ def login(self, form):
+                audit_msg = f"The authenticated user {username} is not" \
+                    " authorized to access pgAdmin based on OAUTH2 config."
+                current_app.logger.warning(audit_msg)
+                return False, gettext(return_msg)
```
gettext should be applied on the string not on a variable.

EvertonSA replied:
hmm, sorry I don't understand, return_msg is a string variable, I just used the same code a few lines above

EvertonSA replied:
i will return the return_msg without gettext

EvertonSA commented:
hi colleagues, I have adjusted according to the comments, does it look better now?

I would like to have a failed check on the test_oauth2_with_mocking.py with all the wrong inputs I have added validation for, but I'm not an experienced person on test/mocking/asserting. If this is mandatory, can someone help me?

khushboovashi commented on web/pgadmin/authenticate/oauth2.py:
```diff
@@ -151,6 +151,26 @@ def login(self, form):
+                                                            additinal_claims)
+
+        if not valid:
+            return_msg = "Your user it's not authorized to access" \
```
return_msg = gettext("The user is not authorized to login "
                                 "based on the claims in the profile. "
                                 "Please contact your administrator.")

EvertonSA commented:
@khushboovashi  thanks a lot for noticing the typos and fixing statements. looks much better now!

ducnm0711 commented:
Hello there @EvertonSA 

I'm testing pgadmin + github org and bounced back into this. Very thankful for all the contributions :)
Currently do we have any way to override the scope config? 

https://github.com/pgadmin-org/pgadmin4/blob/master/web/pgadmin/authenticate/oauth2.py#L118-L119

I tried `'OAUTH2_SCOPE:': 'user:email read:org'`  but scope config doesn't seem to pick up.  It can only fetch user public information, not org.

EvertonSA commented:
> this is how grafana is checking if user is part of organization https://github.com/grafana/grafana/blob/f005a375db149be1981b8fad88a99d34c998cdc9/pkg/login/social/github_oauth.go#L67
> 
> From my humble understanding, this is another issue/PR. 

@ducnm0711 fyi
